src/recursive_shared_mutex.h

```diff
@@ -0,0 +1,101 @@
+ * check
+ * for another write lock already
+ * queued up so we dont jump the line)
+ */
+
+
+static const std::thread::id NON_THREAD_ID = std::thread::id();
+
+class recursive_shared_mutex
+{
+private:
+    // Only locked when accessing counters, ids, or waiting on condition variables.
+    std::mutex _mutex;
+
+    // the read_gate is locked (blocked) when threads have write ownership
+    std::condition_variable _read_gate;
+
+    // the write_gate is locked (blocked) when threads have read ownership
+    std::condition_variable _write_gate;
+
+    // promotion_gates
+    std::condition_variable _promotion_read_gate;
+    std::condition_variable _promotion_write_gate;
+
+    // holds a list of owner ids that have shared ownership and the number of times they locked it
+    std::map<std::thread::id, uint64_t> _read_owner_ids;
+    // holds a list of owner ids that have been auto unlocked due to promoting and the number of
+    // times they were auto unlocked
+    std::map<std::thread::id, uint64_t> _auto_unlocked_ids;
+
+    uint64_t _write_counter;
+    uint64_t _write_promotion_counter;
+    std::thread::id _write_owner_id;
+    std::thread::id _promotion_candidate_id;
+
+private:
+    bool check_for_write_lock(const std::thread::id &locking_thread_id);
+    bool unlock_if_write_lock(const std::thread::id &locking_thread_id);
+
+    void lock_shared_internal(const std::thread::id &locking_thread_id);
+    void lock_shared_internal(const std::thread::id &locking_thread_id, const uint64_t &count);
+    void unlock_shared_internal(const std::thread::id &locking_thread_id);
+    void unlock_shared_internal(const std::thread::id &locking_thread_id, const uint64_t &count);
+    uint64_t get_shared_lock_count(const std::thread::id &locking_thread_id);
+
+    void lock_auto_locks(const std::thread::id &locking_thread_id, const uint64_t &count);
+    uint64_t get_auto_lock_count(const std::thread::id &locking_thread_id);
+    void unlock_auto_locks(const std::thread::id &locking_thread_id);
+
+public:
+    recursive_shared_mutex()
+    {
+        _read_owner_ids.clear();
+        _auto_unlocked_ids.clear();
+        _write_counter = 0;
+        _write_promotion_counter = 0;
+        _write_owner_id = NON_THREAD_ID;
+        _promotion_candidate_id = NON_THREAD_ID;
+    }
+
+    ~recursive_shared_mutex() {}
+    recursive_shared_mutex(const recursive_shared_mutex &) = delete;
+    recursive_shared_mutex &operator=(const recursive_shared_mutex &) = delete;
+
+    void lock(const std::thread::id &locking_thread_id);
+    bool try_promotion(const std::thread::id &locking_thread_id);
+    bool try_lock(const std::thread::id &locking_thread_id);
+    void unlock(const std::thread::id &locking_thread_id);
+    void lock_shared(const std::thread::id &locking_thread_id);
+    bool try_lock_shared(const std::thread::id &locking_thread_id);
+    void unlock_shared(const std::thread::id &locking_thread_id);
```

Comment:
Why pass the thread id?  Shouldn't we access it from the current thread?  Even if there's a good reason that thread A might call these APIs on behalf of thread B, it seems like the most common use will be the "this" thread, so we should have a default parameter that supplies it.

Reply:
It was useful early on for debugging but serves little to no purpose now that it is done. i will remove this and just grab this_thread's id when needed